gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/HttpGremlinResponseStreamDecoder.java

```diff
@@ -0,0 +1,95 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.tinkerpop.gremlin.driver.handler;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.MessageToMessageDecoder;
+import io.netty.handler.codec.http.DefaultHttpContent;
+import io.netty.handler.codec.http.DefaultHttpObject;
+import io.netty.handler.codec.http.DefaultHttpResponse;
+import io.netty.handler.codec.http.DefaultLastHttpContent;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.util.Attribute;
+import io.netty.util.AttributeKey;
+import io.netty.util.AttributeMap;
+import org.apache.tinkerpop.gremlin.util.message.ResponseMessage;
+import org.apache.tinkerpop.gremlin.util.ser.MessageTextSerializerV4;
+import org.apache.tinkerpop.gremlin.util.ser.SerializationException;
+
+import java.util.List;
+import java.util.Objects;
+
+public class HttpGremlinResponseStreamDecoder extends MessageToMessageDecoder<DefaultHttpObject> {
+
+    // todo: move out
+    public static final AttributeKey<Boolean> IS_FIRST_CHUNK = AttributeKey.valueOf("isFirstChunk");
+
+    private final MessageTextSerializerV4<?> serializer;
+
+    public HttpGremlinResponseStreamDecoder(MessageTextSerializerV4<?> serializer) {
+        this.serializer = serializer;
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, DefaultHttpObject msg, List<Object> out) throws Exception {
+        final Attribute<Boolean> isFirstChunk = ((AttributeMap) ctx).attr(IS_FIRST_CHUNK);
+
+        System.out.println("HttpGremlinResponseStreamDecoder start");
+
+        if (msg instanceof DefaultHttpResponse) {
+            isFirstChunk.set(true);
+        }
+
+        if (msg instanceof DefaultHttpContent) {
+            try {
+                if (isFirstChunk.get()) {
+                    System.out.println("first chunk");
+                } else {
+                    System.out.println("not first chunk");
+                }
+
+                final ResponseMessage chunk = serializer.readChunk(((DefaultHttpContent) msg).content(), isFirstChunk.get());
+
+                if (chunk.getResult().getData() != null) {
+                    System.out.println("payload size: " + ((List) chunk.getResult().getData()).size());
+                }
+
+                if (msg instanceof DefaultLastHttpContent) {
+                    final HttpHeaders trailingHeaders = ((DefaultLastHttpContent) msg).trailingHeaders();
+
+                    System.out.println("final chunk, trailing headers:");
+                    System.out.println(trailingHeaders);
+
+                    if (!Objects.equals(trailingHeaders.get("code"), "200")) {
+                        throw new Exception(trailingHeaders.get("message"));
+                    }
+                }
+
+                isFirstChunk.set(false);
+
+                out.add(chunk);
```

Comment:
We might want to make something like a PartialResponseMessage class. It could get a little confusing for others that don't realize that one large ResponseMessage is formed by serializing multiple partial ResponseMessage.